src/qibolab/compilers/compiler.py

```diff
@@ -141,27 +114,31 @@ def compile(self, circuit, platform):
         sequence = PulseSequence()
         # FIXME: This will not work with qubits that have string names
         # TODO: Implement a mapping between circuit qubit ids and platform ``Qubit``s
-        virtual_z_phases = defaultdict(int)
 
         measurement_map = {}
+        qubit_clock = defaultdict(int)
+        channel_clock = defaultdict(int)
         # process circuit gates
-        delays = defaultdict(int)
         for moment in circuit.queue.moments:
-            moment_start = sequence.finish
             for gate in set(filter(lambda x: x is not None, moment)):
                 if isinstance(gate, gates.Align):
                     for qubit in gate.qubits:
-                        delays[qubit] += gate.delay
+                        qubit_clock[qubit] += gate.delay
                     continue
-                gate_sequence, gate_phases = self._compile_gate(
-                    gate, platform, sequence, virtual_z_phases, moment_start, delays
-                )
-                for qubit in gate.qubits:
-                    delays[qubit] = 0
-
-                # update virtual Z phases
-                for qubit, phase in gate_phases.items():
-                    virtual_z_phases[qubit] += phase
+
+                rule = self[gate.__class__]
+                # get local sequence and phases for the current gate
+                gate_sequence = rule(gate, platform)
+                for pulse in gate_sequence:
+                    if qubit_clock[pulse.qubit] > channel_clock[pulse.qubit]:
+                        delay = qubit_clock[pulse.qubit] - channel_clock[pulse.channel]
+                        sequence.append(Delay(delay, pulse.channel))
```

Comment:
> That before #792 means per-qubit

I am not sure if I agree with this, because each qubit may have several logical channels associated to it (readout, drive, flux) and `Delay`s can be applied independently in any of those.

Generally, I hope that #792 will clear these issues but we cannot do everything in the same PR so I tried to do some intermediate solution here. I wanted to avoid putting `qubit` attribute to the `Delay` given that we are planning to drop it from the `Pulse`.